libs/@local/hql/span/src/storage.rs

```diff
@@ -1,18 +1,26 @@
+use alloc::sync::Arc;
+
+use orx_concurrent_vec::ConcurrentVec;
+
 use crate::{tree::SpanNode, Span, SpanId};
 
 /// A collection of spans within a single source.
 ///
 /// The storage is append-only and does not support the removal or modification of spans.
 /// Once inserted spans are to be considered immutable and can be referred to via their `SpanId`,
 /// which is returned on insertion.
+#[derive(Debug)]
 pub struct SpanStorage<S> {
-    inner: Vec<S>,
+    // ConcurrentVec is also used by salsa / nikomatsakis
+    inner: Arc<ConcurrentVec<S>>,
```

Comment:
I'm not a fan of making this an `Arc`. Instead, I think we should make this `!Clone` again and wrap it in an `Arc` where we need. `span_storage.clone()` is misleading as this converts the storage into a smart pointer.

Reply:
I've done this primarily for the parsing (#4828), I had it `!Clone` before, but it made handling the storage quite a bit challenging. (just a lot of lifetimes), and once wrapped, you can't unwrap it.

I can `!Clone` it here, but then would need to make the upstream changes.

Reply:
Why you cannot pass `Arc<SpanStorage<S>>` around when parsing?

Reply:
sure you can, but then you can't remove it out of the `Arc` again, so I thought, that it might just make things easier (although not as visible) to put the `Arc` just into the storage itself.

Reply:
because you're likely to use it in an `Arc` anyway

Reply:
For what scenario you need to remove it from the `Arc` again? It has internal mutability and it's possible to get references to the inner type easily. If functions only require a reference to the storage you possibly even safe indirections and it's more visible what the storage actually is.

Reply:
> For what scenario you need to remove it from the Arc again?

exactly, that is why I put the `Arc` into it, I can - of course - also move it out of storage - tho I didn't want to have duplicate typing for it

Reply:
The semantic meaning of `Clone` on a storage is very different. We have lints to _not_ call `smart_pointer.clone()`, but we don't have that for `SpanStorage`. Further, it is not possible to implement a deep clone on `SpanStorage` anymore. 
Can we please move out the `Arc`? It should be possible to simply replace all usages of `SpanStorage` with `Arc<SpanStorage>` (which leads to the question: Do it needs to be `Arc` or is `Rc` enough?).

Reply:
No, I get that, you're right, I just wanted to avoid writing `Arc<SpanStorage<Span>>` all over the place 😅

Reply:
fixed in [e7d3c7a](https://github.com/hashintel/hash/pull/4826/commits/e7d3c7aab1ee1d9f3de2643e466067de9637837b)

Reply:
I absolutely hear you and see your motivation 🙂 
Thank you!